From f16ca69c1c2635fc85f8301ab7b0c5d3c21b4773 Mon Sep 17 00:00:00 2001
From: Thomas Epperson <thomas.epperson@gmail.com>
Date: Fri, 1 Dec 2017 20:46:07 -0500
Subject: [PATCH 1/3] Fixups to allow compiling with recent gcc
 obj/image-crom.elf fails to link, complaints of missing calloc and
 try_elf_boot

Signed-off-by: Thomas Epperson <thomas.epperson@gmail.com>
---
 Makefile                              | 15 +++++----------
 boot/BootResetAction.c                |  5 +++++
 boot/LoadLinux.c                      |  7 +++++--
 boot_rom/2bBootLibrary.c              |  2 +-
 boot_rom/2bBootStartBios.c            |  2 ++
 boot_rom/2bload.h                     |  3 ++-
 boot_rom/bootrom.ld                   |  2 ++
 drivers/ide/BootIde.c                 |  2 ++
 drivers/pci/BootInterrupts.c          |  2 +-
 drivers/usb/host/ohci-pci.c           |  2 +-
 drivers/usb/sys/BootUSB.c             |  2 ++
 drivers/usb/sys/usbwrapper.c          |  2 ++
 drivers/video/BootVgaInitialization.c |  9 +++++++++
 drivers/video/BootVideoHelpers.c      |  2 ++
 drivers/video/focus.c                 |  2 ++
 etherboot/arch/i386/core/xbox.c       |  5 +++++
 fs/cdrom/iso9660.c                    |  2 +-
 fs/fatx/BootFATX.c                    |  2 +-
 fs/grub/char_io.c                     |  6 +++---
 fs/grub/fsys_ufs2.c                   |  2 ++
 include/boot.h                        |  1 +
 include/stdint.h                      | 25 ++++++-------------------
 lib/crypt/sha1.h                      |  2 +-
 lib/eeprom/BootEEPROM.c               |  2 ++
 lib/misc/BootParser.c                 |  3 +++
 lib/misc/LED.c                        |  3 +++
 lib/misc/vsprintf.c                   |  1 -
 lib/readcd/Makefile                   |  2 +-
 menu/actions/FlashMenuActions.c       |  2 ++
 menu/actions/HddMenuActions.c         |  2 ++
 menu/actions/MenuActions.c            | 13 +++++++++++--
 menu/actions/ResetMenuActions.c       |  4 ++++
 menu/actions/VideoMenuActions.c       |  3 +++
 menu/iconmenu/IconMenuInit.c          |  5 +++++
 menu/misc/Confirm.c                   |  3 +++
 menu/textmenu/FlashMenuInit.c         |  2 ++
 menu/textmenu/HddMenuInit.c           |  2 ++
 menu/textmenu/ResetMenuInit.c         |  2 ++
 menu/textmenu/TextMenuInit.c          |  6 ++++++
 menu/textmenu/VideoMenuInit.c         |  2 ++
 scripts/ldscript-crom.ld              |  2 ++
 setcygwinenv                          |  6 +++---
 tools/Makefile                        |  2 --
 43 files changed, 121 insertions(+), 50 deletions(-)

diff --git a/Makefile b/Makefile
index 80ec92b..9ff2e44 100644
--- a/Makefile
+++ b/Makefile
@@ -1,9 +1,9 @@
-CC	= gcc
+#CC	= gcc
 # prepare check for gcc 3.3, $(GCC_3.3) will either be 0 or 1
 GCC_3.3 := $(shell expr `$(CC) -dumpversion` \>= 3.3)
 
-ETHERBOOT := yes
-INCLUDE = -I$(TOPDIR)/grub -I$(TOPDIR)/include -I$(TOPDIR)/ -I./ -I$(TOPDIR)/fs/cdrom \
+#ETHERBOOT := yes
+INCLUDE += -I$(TOPDIR)/grub -I$(TOPDIR)/include -I$(TOPDIR)/ -I./ -I$(TOPDIR)/fs/cdrom \
 	-I$(TOPDIR)/fs/fatx -I$(TOPDIR)/fs/grub -I$(TOPDIR)/lib/eeprom -I$(TOPDIR)/lib/crypt \
 	-I$(TOPDIR)/drivers/video -I$(TOPDIR)/drivers/ide -I$(TOPDIR)/drivers/flash -I$(TOPDIR)/lib/misc \
 	-I$(TOPDIR)/boot_xbe/ -I$(TOPDIR)/fs/grub -I$(TOPDIR)/lib/font \
@@ -14,18 +14,13 @@ INCLUDE = -I$(TOPDIR)/grub -I$(TOPDIR)/include -I$(TOPDIR)/ -I./ -I$(TOPDIR)/fs/
 CROM_CFLAGS=$(INCLUDE)
 
 #You can override these if you wish.
-CFLAGS= -O2 -g -march=pentium -Werror -pipe -fomit-frame-pointer -Wstrict-prototypes
+CFLAGS= -O2 -g -march=pentium -Werror -fomit-frame-pointer -Wstrict-prototypes -ffreestanding
 
 # add the option for gcc 3.3 only, again, non-overridable
 ifeq ($(GCC_3.3), 1)
 CROM_CFLAGS += -fno-zero-initialized-in-bss
 endif
 
-LD      = ld
-OBJCOPY = objcopy
-
-export CC
-
 TOPDIR  := $(shell /bin/pwd)
 SUBDIRS	= boot_rom fs drivers lib boot menu
 #### Etherboot specific stuff
@@ -33,7 +28,7 @@ ifeq ($(ETHERBOOT), yes)
 ETH_SUBDIRS = etherboot
 CROM_CFLAGS	+= -DETHERBOOT
 ETH_INCLUDE = 	-I$(TOPDIR)/etherboot/include -I$(TOPDIR)/etherboot/arch/i386/include	
-ETH_CFLAGS  = 	-O2 -mcpu=pentium -Werror $(ETH_INCLUDE) -Wstrict-prototypes -fomit-frame-pointer -pipe -Ui386
+ETH_CFLAGS  = 	-O2 -march=pentium -mtune=pentium -Werror $(ETH_INCLUDE) -Wstrict-prototypes -fomit-frame-pointer -pipe -Ui386
 endif
 
 LDFLAGS-ROM     = -s -S -T $(TOPDIR)/scripts/ldscript-crom.ld
diff --git a/boot/BootResetAction.c b/boot/BootResetAction.c
index ff59cef..2fa14d6 100644
--- a/boot/BootResetAction.c
+++ b/boot/BootResetAction.c
@@ -22,6 +22,11 @@
 #include "video.h"
 #include "memory_layout.h"
 
+extern char *VideoEncoderName(void);
+extern char *AvCableName(void);
+extern void IconMenuInit(void);
+extern void IconMenu(void);
+
 JPEG jpegBackdrop;
 
 int nTempCursorMbrX, nTempCursorMbrY;
diff --git a/boot/LoadLinux.c b/boot/LoadLinux.c
index a06d971..56f697e 100644
--- a/boot/LoadLinux.c
+++ b/boot/LoadLinux.c
@@ -37,8 +37,11 @@ void startLinux(void* initrdStart, unsigned long initrdSize, const char* appendL
 void setup(void* KernelPos, void* PhysInitrdPos, unsigned long InitrdSize, const char* kernel_cmdline);
 void I2CRebootSlow(void);
 
-void try_elf_boot (char* data, int len);
-
+//try_elf_boot is part of etherboot, which is not being compiled
+//TODO: add compile time check for this
+void try_elf_boot (char* data, int len)
+{
+}
 
 void BootPrintConfig(CONFIGENTRY *config) {
 	int CharsProcessed=0, CharsSinceNewline=0, Length=0;
diff --git a/boot_rom/2bBootLibrary.c b/boot_rom/2bBootLibrary.c
index 92a3076..424983c 100644
--- a/boot_rom/2bBootLibrary.c
+++ b/boot_rom/2bBootLibrary.c
@@ -10,7 +10,7 @@
 */
 // 20040924 - Updated by dmp to include more str functions, and use ASM
 // where possible. ASM shamelessly stolen from linux-2.6.8.1
-
+typedef unsigned int size_t;
 #include "stdint.h"
 
 size_t strlen(const char * s)
diff --git a/boot_rom/2bBootStartBios.c b/boot_rom/2bBootStartBios.c
index e35a643..9f09c07 100644
--- a/boot_rom/2bBootStartBios.c
+++ b/boot_rom/2bBootStartBios.c
@@ -16,6 +16,8 @@
 #include "2bload.h"
 #include "sha1.h"
 
+extern void setLED(void *pattern);
+
 extern int decompress_kernel(char*out, char *data, int len);
 
 u32 PciWriteDword(unsigned int bus, unsigned int dev, unsigned int func, unsigned int reg_off, unsigned int dw) 
diff --git a/boot_rom/2bload.h b/boot_rom/2bload.h
index 61e4e0f..26d535e 100644
--- a/boot_rom/2bload.h
+++ b/boot_rom/2bload.h
@@ -10,11 +10,12 @@
  *                                                                         *
  ***************************************************************************/
 
+typedef unsigned int size_t;
+
 #include "2bconsts.h"
 #include "stdint.h"
 #include "cromwell_types.h"
 
-
 /////////////////////////////////
 // LED-flashing codes
 // or these together as argument to I2cSetFrontpanelLed
diff --git a/boot_rom/bootrom.ld b/boot_rom/bootrom.ld
index 82b8775..cd87ebf 100644
--- a/boot_rom/bootrom.ld
+++ b/boot_rom/bootrom.ld
@@ -32,6 +32,8 @@ SECTIONS {
 		_ram_location = .;
 		_start_ramcopy = _end_rom;
 		*(.text);
+		*(.text.unlikely);
+		*(.eh_frame);
 		_start_checksum = _start_ramcopy - LOW_ROM;
 	}
 
diff --git a/drivers/ide/BootIde.c b/drivers/ide/BootIde.c
index 41e2a2d..dcf7034 100644
--- a/drivers/ide/BootIde.c
+++ b/drivers/ide/BootIde.c
@@ -11,6 +11,8 @@
 
 #undef sprintf
 
+extern int VideoDumpAddressAndData(u32 dwAds, const u8 *baData, u32 dwCountBytesUsable);
+
 ////////////////////////////////////
 // IDE types and constants
 #define IDE_SECTOR_SIZE 		0x200
diff --git a/drivers/pci/BootInterrupts.c b/drivers/pci/BootInterrupts.c
index 07b4c75..460942c 100644
--- a/drivers/pci/BootInterrupts.c
+++ b/drivers/pci/BootInterrupts.c
@@ -211,7 +211,7 @@ void BootInterruptsWriteIdt() {
 			ptspmi[n].m_wHandlerLinearAddressHigh16=(u16)(((u32)isrprep[n1].m_dwpVector)>>16);
 			n1++;
 		} else { // otherwise default handler (pretty useless, but will catch it)
-			ptspmi[n].m_wHandlerHighAddressLow16=(u16)IntHandlerUnused;
+			ptspmi[n].m_wHandlerHighAddressLow16=(u16)((u32)IntHandlerUnused);
 			ptspmi[n].m_wHandlerLinearAddressHigh16=(u16)(((u32)IntHandlerUnused)>>16);
 		}
 	}
diff --git a/drivers/usb/host/ohci-pci.c b/drivers/usb/host/ohci-pci.c
index 911a52f..7117a6d 100755
--- a/drivers/usb/host/ohci-pci.c
+++ b/drivers/usb/host/ohci-pci.c
@@ -350,7 +350,7 @@ static const struct hc_driver ohci_pci_hc_driver = {
 
 /*-------------------------------------------------------------------------*/
 
-static const struct pci_device_id __devinitdata pci_ids [] = { {
+static struct pci_device_id __devinitdata pci_ids [] = { {
 
 	/* handle any USB OHCI controller */
 	.class =	(PCI_CLASS_SERIAL_USB << 8) | 0x10,
diff --git a/drivers/usb/sys/BootUSB.c b/drivers/usb/sys/BootUSB.c
index 243d101..0b21fa8 100755
--- a/drivers/usb/sys/BootUSB.c
+++ b/drivers/usb/sys/BootUSB.c
@@ -11,6 +11,8 @@ void subsys_usb_init(void);
 void module_exit_usb_exit(void);
 
 extern struct pci_device_id  *module_table_pci_ids;
+extern void UsbKeyBoardInit(void);
+extern void UsbKeyBoardRemove(void);
 
 // straigth call...
 int usb_hcd_pci_probe (struct pci_dev *dev, const struct pci_device_id *id);
diff --git a/drivers/usb/sys/usbwrapper.c b/drivers/usb/sys/usbwrapper.c
index ce952e4..7be1203 100755
--- a/drivers/usb/sys/usbwrapper.c
+++ b/drivers/usb/sys/usbwrapper.c
@@ -9,6 +9,8 @@
 #include <stdarg.h>
 #include "video.h"
 
+extern int vsprintf(char *buf, const char *fmt, va_list args);
+
 /*------------------------------------------------------------------------*/ 
 // Output window for USB messages
 int usb_curs_x=0;
diff --git a/drivers/video/BootVgaInitialization.c b/drivers/video/BootVgaInitialization.c
index 3133512..e6422b0 100644
--- a/drivers/video/BootVgaInitialization.c
+++ b/drivers/video/BootVgaInitialization.c
@@ -27,6 +27,15 @@
 #include "encoder.h"
 #include "xcalibur.h"
 
+extern int conexant_calc_hdtv_mode(xbox_hdtv_mode hdtv_mode, unsigned char pll_int, void **mode_out);
+extern int conexant_calc_vga_mode(xbox_av_type av_type, unsigned char pll_int, void **mode_out);
+extern int conexant_calc_mode(xbox_video_mode * mode, struct riva_regs * riva_out);
+extern int focus_calc_hdtv_mode(xbox_hdtv_mode hdtv_mode, unsigned char pll_int, void **encoder_regs);
+extern int focus_calc_mode(xbox_video_mode * mode, struct riva_regs * riva_out );
+extern int I2CWriteBytetoRegister(u8 bPicAddressI2cFormat, u8 bRegister, u8 wDataToWrite);
+extern int ReadfromSMBus(u8 Address,u8 bRegister,u8 Size,u32 *Data_to_smbus);
+extern int WriteToSMBus(u8 Address,u8 bRegister,u8 Size,u32 Data_to_smbus);
+
 void DetectVideoEncoder(void) {
 	if (I2CTransmitByteGetReturn(0x45,0x00) != ERR_I2C_ERROR_BUS) video_encoder = ENCODER_CONEXANT;
 	else if (I2CTransmitByteGetReturn(0x6a,0x00) != ERR_I2C_ERROR_BUS) video_encoder = ENCODER_FOCUS;
diff --git a/drivers/video/BootVideoHelpers.c b/drivers/video/BootVideoHelpers.c
index 71a3f33..47a04b5 100644
--- a/drivers/video/BootVideoHelpers.c
+++ b/drivers/video/BootVideoHelpers.c
@@ -22,6 +22,8 @@
 #include "decode-jpg.h"
 #define WIDTH_SPACE_PIXELS 5
 
+extern int vsprintf(char *buf, const char *fmt, va_list args);
+
 // returns number of x pixels taken up by ascii character bCharacter
 
 unsigned int BootVideoGetCharacterWidth(u8 bCharacter, bool fDouble)
diff --git a/drivers/video/focus.c b/drivers/video/focus.c
index 0ddf82d..f76c03f 100644
--- a/drivers/video/focus.c
+++ b/drivers/video/focus.c
@@ -23,6 +23,8 @@ typedef struct _focus_pll_settings{
 	int tv_vtotal;
 } focus_pll_settings;
 
+int focus_calc_pll_settings(focus_pll_settings *settings, char *regs);
+
 static const unsigned char focus_defaults[0xc4] = {
 	/*0x00*/ 0x00,0x00,0x00,0x00,0x80,0x02,0xaa,0x0a,
 	/*0x08*/ 0x00,0x10,0x00,0x00,0x03,0x21,0x15,0x04,
diff --git a/etherboot/arch/i386/core/xbox.c b/etherboot/arch/i386/core/xbox.c
index e1dc2a8..cf2b40e 100644
--- a/etherboot/arch/i386/core/xbox.c
+++ b/etherboot/arch/i386/core/xbox.c
@@ -7,6 +7,11 @@
 
 #include <stdarg.h>
 
+typedef unsigned int u32;
+
+extern int printk(const char *szFormat, ...);
+extern void wait_us(u32 ticks);
+
 #define MAX_APPEND_LINE_LENGTH 512
 
 extern struct pci_driver forcedeth_driver;
diff --git a/fs/cdrom/iso9660.c b/fs/cdrom/iso9660.c
index 0ce326b..3a91fac 100644
--- a/fs/cdrom/iso9660.c
+++ b/fs/cdrom/iso9660.c
@@ -22,7 +22,7 @@ int isupper( int ch )
 	return (unsigned int)(ch - 'A') < 26u;
 }
 
-strip_blank(char *buffer, unsigned char len) {
+void strip_blank(char *buffer, unsigned char len) {
 	unsigned char i;
 	for(i = len; i > 0; i--) {
 		if(buffer[i] != ' ') {
diff --git a/fs/fatx/BootFATX.c b/fs/fatx/BootFATX.c
index 736316a..1c61f8e 100644
--- a/fs/fatx/BootFATX.c
+++ b/fs/fatx/BootFATX.c
@@ -5,10 +5,10 @@
 #include "BootFATX.h"
 #include <sys/types.h>
 
-
 #undef FATX_DEBUG
 
 //#define FATX_INFO
+extern int strncmp(const char * cs,const char * ct,size_t count);
 
 int checkForLastDirectoryEntry(unsigned char* entry) {
 
diff --git a/fs/grub/char_io.c b/fs/grub/char_io.c
index 9ce8f96..18078db 100644
--- a/fs/grub/char_io.c
+++ b/fs/grub/char_io.c
@@ -19,6 +19,7 @@
  */
 
 #include <shared.h>
+extern int tolower(int ch);
 
 char *
 convert_to_ascii (char *buf, int c,...)
@@ -128,9 +129,8 @@ grub_isspace (int c)
    a static library supporting minimal standard C functions and link
    each image with the library. Complicated things should be left to
    computer, definitely. -okuji  */
-#if !defined(STAGE1_5) || defined(FSYS_VSTAFS)
-int
-grub_strcmp (const char *s1, const char *s2)
+#if 1 //!defined(STAGE1_5) || defined(FSYS_VSTAFS)
+int grub_strcmp (const char *s1, const char *s2)
 {
   while (*s1 || *s2)
     {
diff --git a/fs/grub/fsys_ufs2.c b/fs/grub/fsys_ufs2.c
index 1971675..49b6c4a 100644
--- a/fs/grub/fsys_ufs2.c
+++ b/fs/grub/fsys_ufs2.c
@@ -61,6 +61,8 @@
 
 #include "ufs2.h"
 
+#define isspace grub_isspace
+
 /* used for filesystem map blocks */
 static int mapblock;
 static int mapblock_offset;
diff --git a/include/boot.h b/include/boot.h
index c2cff4e..c755ff5 100644
--- a/include/boot.h
+++ b/include/boot.h
@@ -1,6 +1,7 @@
 #ifndef _Boot_H_
 #define _Boot_H_
 
+typedef unsigned int size_t;
 #include "config.h"
 
 /***************************************************************************
diff --git a/include/stdint.h b/include/stdint.h
index 89edf8d..c6129be 100644
--- a/include/stdint.h
+++ b/include/stdint.h
@@ -1,19 +1,6 @@
-#ifndef STDINT_H
-#define STDINT_H
-
-typedef unsigned           size_t;
-
-typedef unsigned char      uint8_t;
-typedef unsigned short     uint16_t;
-typedef unsigned long      uint32_t;
-typedef unsigned long long uint64_t;
-
-typedef signed char        int8_t;
-typedef signed short       int16_t;
-typedef signed long        int32_t;
-typedef signed long long   int64_t;
-
-typedef short int          int_least16_t;
-typedef unsigned short int uint_least16_t;
-
-#endif /* STDINT_H */
+#ifndef __STDINT_H_
+#define __STDINT_H_
+typedef unsigned int uint32_t;
+typedef unsigned char uint8_t;
+typedef short int_least16_t;
+#endif
diff --git a/lib/crypt/sha1.h b/lib/crypt/sha1.h
index e2b9f75..4dc9e94 100644
--- a/lib/crypt/sha1.h
+++ b/lib/crypt/sha1.h
@@ -17,7 +17,7 @@
 #ifndef _SHA1_H_
 #define _SHA1_H_  
 
-#include <stdint.h>
+#include "stdint.h"
 /*
  * If you do not have the ISO standard stdint.h header file, then you
  * must typdef the following:
diff --git a/lib/eeprom/BootEEPROM.c b/lib/eeprom/BootEEPROM.c
index a0f3545..81c86ef 100644
--- a/lib/eeprom/BootEEPROM.c
+++ b/lib/eeprom/BootEEPROM.c
@@ -2,6 +2,8 @@
 #include "VideoInitialization.h"
 #include "BootEEPROM.h"
 
+extern int WriteToSMBus(u8 Address,u8 bRegister,u8 Size,u32 Data_to_smbus);
+
 void BootEepromReadEntireEEPROM() {
 	int i;
 	u8 *pb=(u8 *)&eeprom;
diff --git a/lib/misc/BootParser.c b/lib/misc/BootParser.c
index e5c7e33..9ffcefd 100644
--- a/lib/misc/BootParser.c
+++ b/lib/misc/BootParser.c
@@ -1,5 +1,8 @@
 #include "boot.h"
 
+extern char * strsep(char **s, const char *ct);
+extern int strncmp(const char * cs,const char * ct,size_t count);
+
 CONFIGENTRY *ParseConfig(char *szBuffer, unsigned int fileLen, char *szPath) {
 	char *linePtr;
 	char *currentPos = szBuffer;
diff --git a/lib/misc/LED.c b/lib/misc/LED.c
index 151523b..581ea76 100644
--- a/lib/misc/LED.c
+++ b/lib/misc/LED.c
@@ -17,6 +17,9 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
 
+typedef unsigned char u8;
+extern int I2cSetFrontpanelLed(u8 b);
+
 // Set the pattern of the LED.
 // The pattern must be 4 characters long and must consist
 // only of 'r', 'g', 'o' and 'x'.
diff --git a/lib/misc/vsprintf.c b/lib/misc/vsprintf.c
index 0ab7c7f..c2b86fd 100644
--- a/lib/misc/vsprintf.c
+++ b/lib/misc/vsprintf.c
@@ -12,7 +12,6 @@
 
 #include <stdarg.h>
 #include <sys/types.h>
-#include <string.h>
 
 /* haha, don't need ctype.c */
 #define isdigit(c)	((c) >= '0' && (c) <= '9')
diff --git a/lib/readcd/Makefile b/lib/readcd/Makefile
index 22136db..19574b1 100644
--- a/lib/readcd/Makefile
+++ b/lib/readcd/Makefile
@@ -4,4 +4,4 @@ clean:
 	rm -f readcd
 
 readcd:
-	gcc -DDEBUG_ISO -DSTANDALONE -I./ -I ../../fs/cdrom -o readcd main.c ../../fs/cdrom/iso9660.c
+	$(CC) -DDEBUG_ISO -DSTANDALONE -I./ -I ../../fs/cdrom -o readcd main.c ../../fs/cdrom/iso9660.c
diff --git a/menu/actions/FlashMenuActions.c b/menu/actions/FlashMenuActions.c
index 232593a..2fcbecd 100644
--- a/menu/actions/FlashMenuActions.c
+++ b/menu/actions/FlashMenuActions.c
@@ -8,6 +8,8 @@
  ***************************************************************************/
 #include "FlashMenuActions.h"
 
+extern int BootLoadFlashCD(int cdromId);
+
 void FlashBiosFromCD(void *cdromId) {
 #ifdef FLASH
 	BootLoadFlashCD(*(int *)cdromId);
diff --git a/menu/actions/HddMenuActions.c b/menu/actions/HddMenuActions.c
index 27fbc71..13aeb98 100644
--- a/menu/actions/HddMenuActions.c
+++ b/menu/actions/HddMenuActions.c
@@ -8,6 +8,8 @@
  ***************************************************************************/
 #include "HddMenuActions.h"
 
+extern int Confirm(char *message, char *noText, char *yesText, int defaultItem);
+
 void LockHdd(void *driveId) {
 	int nIndexDrive = *(int *)driveId;
 	u8 password[20];
diff --git a/menu/actions/MenuActions.c b/menu/actions/MenuActions.c
index 4d77efc..746efe1 100644
--- a/menu/actions/MenuActions.c
+++ b/menu/actions/MenuActions.c
@@ -20,6 +20,15 @@
 #include "cpu.h"
 #include "VideoInitialization.h"
 #include "TextMenu.h"
+
+extern void TextMenu(TEXTMENU *menu, TEXTMENUITEM *selectedItem);
+extern void TextMenuAddItem(TEXTMENU *menu, TEXTMENUITEM *newMenuItem);
+extern int LoadKernelCdrom(CONFIGENTRY *config);
+extern int LoadKernelFatX(CONFIGENTRY *config);
+extern int LoadKernelNative(CONFIGENTRY *config);
+extern void ExittoLinux(CONFIGENTRY *config);
+extern int BootLoadFlashCD(int cdromId);
+
 CONFIGENTRY *LoadConfigCD(int);
 TEXTMENU *TextMenuInit(void);
 
@@ -112,7 +121,7 @@ void BootMenuEntry(void *entry) {
 }
 
 void DrawChildTextMenu(void *menu) {
-	TextMenu((TEXTMENU*)menu);
+	TextMenu((TEXTMENU*)menu, 0);
 }
 
 #ifdef ETHERBOOT 
@@ -124,7 +133,7 @@ void BootFromEtherboot(void *data) {
 
 #ifdef FLASH
 void FlashBios(void *data) {
-	BootLoadFlashCD();
+	BootLoadFlashCD(*(int*)data);
 }
 #endif
 
diff --git a/menu/actions/ResetMenuActions.c b/menu/actions/ResetMenuActions.c
index 0b54e44..a9db30b 100644
--- a/menu/actions/ResetMenuActions.c
+++ b/menu/actions/ResetMenuActions.c
@@ -8,6 +8,10 @@
  ***************************************************************************/
 #include "ResetMenuActions.h"
 
+extern void I2CRebootSlow(void);
+extern void I2CRebootQuick(void);
+extern void I2CPowerOff(void);
+
 void SlowReboot(void *ignored){
 	I2CRebootSlow();
 }
diff --git a/menu/actions/VideoMenuActions.c b/menu/actions/VideoMenuActions.c
index 2e6d7e2..dc6f49b 100644
--- a/menu/actions/VideoMenuActions.c
+++ b/menu/actions/VideoMenuActions.c
@@ -13,6 +13,9 @@
 #include "VideoInitialization.h"
 #include "BootEEPROM.h"
 
+#define strcmp grub_strcmp
+extern int grub_strcmp (const char *s1, const char *s2);
+
 void SetWidescreen(void *menuItemText) {
 	char *text = (char *)menuItemText;
 	if (!strcmp(text, "Display Size: Widescreen")) {
diff --git a/menu/iconmenu/IconMenuInit.c b/menu/iconmenu/IconMenuInit.c
index ad7bdfe..518e0b9 100644
--- a/menu/iconmenu/IconMenuInit.c
+++ b/menu/iconmenu/IconMenuInit.c
@@ -15,6 +15,11 @@
 #include "IconMenu.h"
 #include "MenuActions.h"
 
+extern void *TextMenuInit(void);
+extern CONFIGENTRY* LoadConfigFatX(void);
+extern CONFIGENTRY* LoadConfigNative(int drive, int partition);
+extern int strncmp(const char * cs,const char * ct,size_t count);
+
 void InitFatXIcons(void);
 void InitNativeIcons(void);
 
diff --git a/menu/misc/Confirm.c b/menu/misc/Confirm.c
index 0a673da..bdb76ad 100644
--- a/menu/misc/Confirm.c
+++ b/menu/misc/Confirm.c
@@ -5,6 +5,9 @@ int Confirm_Result;
 static void Confirm_Yes(void*);
 static void Confirm_No(void*);
 
+extern void TextMenuAddItem(TEXTMENU *menu, TEXTMENUITEM *newMenuItem);
+extern void TextMenu(TEXTMENU *menu, TEXTMENUITEM *selectedItem);
+
 int Confirm(char *message, char *yesText, char *noText, int defaultItem) {
 	TEXTMENU ConfirmMenu;
 	TEXTMENUITEM YesItem, NoItem;
diff --git a/menu/textmenu/FlashMenuInit.c b/menu/textmenu/FlashMenuInit.c
index 1ee369e..418c28a 100644
--- a/menu/textmenu/FlashMenuInit.c
+++ b/menu/textmenu/FlashMenuInit.c
@@ -12,6 +12,8 @@
 #include "TextMenu.h"
 #include "FlashMenuActions.h"
 
+extern void TextMenuAddItem(TEXTMENU *menu, TEXTMENUITEM *newMenuItem);
+
 TEXTMENU* FlashMenuInit(void) {
 	TEXTMENUITEM *itemPtr;
 	TEXTMENU *menuPtr;
diff --git a/menu/textmenu/HddMenuInit.c b/menu/textmenu/HddMenuInit.c
index daec028..8d90253 100644
--- a/menu/textmenu/HddMenuInit.c
+++ b/menu/textmenu/HddMenuInit.c
@@ -12,6 +12,8 @@
 #include "TextMenu.h"
 #include "HddMenuActions.h"
 
+extern void TextMenuAddItem(TEXTMENU *menu, TEXTMENUITEM *newMenuItem);
+
 TEXTMENU *HddMenuInit(void) {
 	TEXTMENUITEM *itemPtr;
 	TEXTMENU *menuPtr;
diff --git a/menu/textmenu/ResetMenuInit.c b/menu/textmenu/ResetMenuInit.c
index e98eabe..acc0db3 100644
--- a/menu/textmenu/ResetMenuInit.c
+++ b/menu/textmenu/ResetMenuInit.c
@@ -11,6 +11,8 @@
 #include "TextMenu.h"
 #include "ResetMenuActions.h"
 
+extern void TextMenuAddItem(TEXTMENU *menu, TEXTMENUITEM *newMenuItem);
+
 TEXTMENU* ResetMenuInit(void) {
 	TEXTMENUITEM *itemPtr;
 	TEXTMENU *menuPtr;
diff --git a/menu/textmenu/TextMenuInit.c b/menu/textmenu/TextMenuInit.c
index b47b098..a568571 100644
--- a/menu/textmenu/TextMenuInit.c
+++ b/menu/textmenu/TextMenuInit.c
@@ -12,6 +12,12 @@
 
 #include "VideoInitialization.h"
 
+extern TEXTMENU *VideoMenuInit(void);
+extern TEXTMENU* ResetMenuInit(void);
+extern TEXTMENU* FlashMenuInit(void);
+extern TEXTMENU* HddMenuInit(void);
+extern void TextMenuAddItem(TEXTMENU *menu, TEXTMENUITEM *newMenuItem);
+
 TEXTMENU *TextMenuInit(void) {
 	
 	TEXTMENUITEM *itemPtr;
diff --git a/menu/textmenu/VideoMenuInit.c b/menu/textmenu/VideoMenuInit.c
index 4cda48c..fb3e545 100644
--- a/menu/textmenu/VideoMenuInit.c
+++ b/menu/textmenu/VideoMenuInit.c
@@ -12,6 +12,8 @@
 
 #include "VideoInitialization.h"
 
+extern void TextMenuAddItem(TEXTMENU *menu, TEXTMENUITEM *newMenuItem);
+
 TEXTMENU *VideoMenuInit(void) {
 	TEXTMENUITEM *itemPtr;
 	TEXTMENU *menuPtr;
diff --git a/scripts/ldscript-crom.ld b/scripts/ldscript-crom.ld
index 01c96cc..19aff28 100644
--- a/scripts/ldscript-crom.ld
+++ b/scripts/ldscript-crom.ld
@@ -22,6 +22,8 @@ SECTIONS {
 	.text LOW_ROM : AT ( 0 ){
 		_start_low_rom = . ;
 		*(.text);
+		*(.text.unlikely);
+		*(.eh_frame);
 		_end_low_rom = . ;
 	}
 
diff --git a/setcygwinenv b/setcygwinenv
index b6e7144..8023ead 100644
--- a/setcygwinenv
+++ b/setcygwinenv
@@ -10,8 +10,8 @@
 # In order for the exports to be global, this script should be run
 # with the following command line: ". setcygwinenv"
 #
-export OBJCOPY=i686-pc-linux-gnu-objcopy
-export LD=i686-pc-linux-gnu-ld
-export CC=i686-pc-linux-gnu-gcc
+#export OBJCOPY=i686-pc-linux-gnu-objcopy
+#export LD=i686-pc-linux-gnu-ld
+#export CC=i686-pc-linux-gnu-gcc
 export -n INCLUDE
 alias make='make -e'
diff --git a/tools/Makefile b/tools/Makefile
index 89941c3..031948e 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -1,4 +1,2 @@
-CC := gcc
-
 all:
 	${CC} -pedantic -Wall -o xdecode xdecode.c
-- 
1.9.1

